### DISCUSSION TITLE
Adding RadialGradientBrush in uno library

### DISCUSSION THREAD

**Discussion by AghaKhan** [unoplatform/uno#6902] [Q&A]
### Current behavior

My Uno library works fine until I add RadialGradientBrush to my project. To resolve this problem, I installed Microsoft.UI.Xaml package, which broke all the packages monoandroid10.0, MonoAndroid11.0, xamarinios10, and xamarinmac20. What is the solution to add RadialGradientBrush, and should it work on all platforms?  
I am Microsoft Visual Studio Professional 2019 Version 16.11.2

### Expected behavior

It should work on all platforms.

### How to reproduce it (as minimally and precisely as possible)

I visited https://playground.platform.uno and add RadialGradientBrush  as a Resource, but unable to find 
[SimpleExample](https://1drv.ms/u/s!AuknqQF1kFXvmjb6rfL9wmrXwD61?e=8A4Exv)

**Comment by AghaKhan** [comment]
[SimpleExample](https://1drv.ms/u/s!AuknqQF1kFXvmjb6rfL9wmrXwD61?e=8A4Exv)

**Comment by jeromelaban** [comment]
`RadialGradientBrush` is part of a different namespace, you'll need to add an additional declaration in your XAML file.

Here's an example from [the Microsoft documentation](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.media.radialgradientbrush?view=winui-3.0#examples):

```xaml
<Page
  xmlns:media="using:Microsoft.UI.Xaml.Media">

  <Rectangle Width="200" Height="200">
      <Rectangle.Fill>
          <media:RadialGradientBrush>
              <GradientStop Color="Blue" Offset="0.0" />
              <GradientStop Color="Yellow" Offset="0.2" />
              <GradientStop Color="LimeGreen" Offset="0.4" />
              <GradientStop Color="LightBlue" Offset="0.6" />
              <GradientStop Color="Blue" Offset="0.8" />
              <GradientStop Color="LightGray" Offset="1" />
          </media:RadialGradientBrush>
      </Rectangle.Fill>
  </Rectangle>
</Page>
```

**Reply by AghaKhan** [reply]
Before posting this question I know that solution and it fails all the time. Build All faild. 
[BugInRadioialGradientBrush](https://1drv.ms/u/s!AuknqQF1kFXvmjiLd5Pp0Hf29AGv?e=S4ER7k)

**Reply by jeromelaban** [reply]
In the sample you posted, the xmlns is incorrectly formatted (`clr-namespace` should not be there). Please see the example I posted above.

**Reply by AghaKhan** [reply]
Thanks for reply. I am not sure what I am missing. Kindly look this image. Now I used your code and still it does not work.
[YourCode](https://1drv.ms/u/s!AuknqQF1kFXvmjmrXaNyGf48yfZC?e=GZIqrU)
Best regards
Agha

**Reply by jeromelaban** [reply]
Thanks. As you are creating a library, you also need to include the WinUI nuget package. Alongside the changes for the `xmlns`, also need to add the following:

```xml
<ItemGroup Condition="'$(TargetFramework)'=='uap10.0.18362'">
  <PackageReference Include="Microsoft.UI.Xaml" Version="2.5.0" />
</ItemGroup>
```

To get the reference to WinUI 2.5 which contains the RadiaGradient brush for uap targets.

**Reply by AghaKhan** [reply]
I have no words to say thank you. It works. Regards Agha

**Reply by AghaKhan** [reply]
Wow. It worked before compiling but failed the project. [Errors](https://1drv.ms/u/s!AuknqQF1kFXvmjb6rfL9wmrXwD61?e=bOrrKu) . Kindly look Errors.txt and errors.jpg. Now in my references, there are warnings, and nothing build. Please note I have created same project in WPF, and it works fine. No issue found. In my opinion Win2D is not part of Android, or iOS, it shouldn't work, but I might be wrong. Regards Agha

**Reply by AghaKhan** [reply]
<Project Sdk="MSBuild.Sdk.Extras/3.0.23">
  
  <PropertyGroup>
    <TargetFrameworks>uap10.0.18362;netstandard2.0;xamarinios10;xamarinmac20;MonoAndroid11.0;monoandroid10.0</TargetFrameworks>
    
    <GenerateLibraryLayout>true</GenerateLibraryLayout>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.UI.Xaml" Version="2.6.2" />
    <PackageReference Include="Uno.UI" Version="3.9.7" />
  </ItemGroup>
  <ItemGroup>
    <Page Include="**\*.xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
    <Compile Update="**\*.xaml.cs">
      <DependentUpon>%(Filename)</DependentUpon>
    </Compile>
  </ItemGroup>
  <ItemGroup>
    <UpToDateCheckInput Include="**\*.xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
  </ItemGroup>
  <ItemGroup>
    <None Remove="UnoWatchCtrl.xaml" />
  </ItemGroup>
  <ItemGroup>
    <UpToDateCheckInput Remove="UnoWatchCtrl.xaml" />
  </ItemGroup>
  <ItemGroup>
    <Page Update="UnoWatchCtrl.xaml">
      <SubType>Designer</SubType>
    </Page>
  </ItemGroup>
</Project>

**Reply by AghaKhan** [reply]
When I tried to add  (TargetFramework is not defined. Also adding <PackageReference Include="Microsoft.UI.Xaml" Version="2.6.2" />
<ItemGroup Condition="'$(TargetFramework)'=='uap10.0.18362'">
  <PackageReference Include="Microsoft.UI.Xaml" Version="2.5.0" />
</ItemGroup>

**Reply by jeromelaban** [reply]
Win2D is supported in Uno, indeed, which is why you are seeing these errors. You can still use it for `uap10.0.xxx` using `<ItemGroup Condition="'$(TargetFramework)'=='uap10.0.18362'">`, but for Uno targets you'll need to use something else like [SkiaSharp](https://github.com/mono/SkiaSharp). You can find a [sample here](https://github.com/unoplatform/Uno.Samples/tree/master/UI/SkiaSharpTest).

By the way, you can paste your images directly in the GitHub text editor, same for the zip files for sources. This way, it will stay with the issue and not in your one drive, if someone looks for this thread after a while.

**Reply by AghaKhan** [reply]
I have downloaded SkiaSharp source code and searched RadialGradientBrush, which I didn't find, so SkiaSharp will not help to resolve the problem of RadialGradientBrush. I don't want to add xmlns:media="using:Microsoft.UI.Xaml.Media" because it is only for UWP and unsuitable for rest of platforms. Can you provide me a sample (draw a circle with RadialGradientBrush) which is good for all platforms?  Best regards Agha

**Reply by jeromelaban** [reply]
`RadialGradientBrush` is supported on all platforms. Can you explain why you need to add Win2D ?

**Comment by AghaKhan** [comment]
The objective of using Uno Platform is to work across all Platforms with one code. If we are only talking about UWP, then it doesn’t make any sense to use Uno Platform. I have written many applications (UWP) where I am using Win2D and without Win2D namespace support of RadialGradientBrush is absent. Microsoft wants to bring Win2D with UWP because RadialGradientBrush is available natively. I just realized if I add Win2D none of Platform will work. I requested a sample (A circle or rectangle) where RadialGradientBrush is available across all Platforms. Indeed, if RadialGradientBrush is available across all Platforms then it must be one line code. Can you provide few lines of code about RadialGradientBrush which should work all Platforms?

**Reply by jeromelaban** [reply]
Can you update your sample with all the adjustments made in this thread? It looks like there's some misunderstanding in what should be done.

**Comment by AghaKhan** [comment]
[OctagonApp.zip](https://github.com/unoplatform/uno/files/7080261/OctagonApp.zip)
Kindly look .\OctagonApp\ControlLib\OctagonControl.xaml
If you are using RadialGradientBrush in your code then you might Microsoft provided code. This code works fine if you add win2D to your project because RadialGradientBrush is defined inside Win2D
xmlns:media="using:Microsoft.UI.Xaml.Media">

Your application will not compile for any Platform if you add Win2D in your application.
Any help will be very much appreciated. 
Best regards Agha Khan

**Reply by jeromelaban** [reply]
Thanks. As I mentioned in https://github.com/unoplatform/uno/discussions/6902#discussioncomment-1251020, you need this reference:
```xml
<ItemGroup Condition="'$(TargetFramework)'=='uap10.0.18362'">
  <PackageReference Include="Microsoft.UI.Xaml" Version="2.5.0" />
</ItemGroup>
```
in your `controllib.csproj`. Otherwise, `RadialGradientBrush` won't be available for `uap10.0.x` as it is defined in the `Microsoft.UI.Xaml` nuget package, as well as add the `media:xmlns` at the top.

Please see in the attached archive for the complete changes:  [6902-ControlLib.zip](https://github.com/unoplatform/uno/files/7080326/6902-ControlLib.zip)

**Reply by AghaKhan** [reply]
Indeed, it worked, but without your help how someone can resolve those problem. Regards Agha
